kvserver: lower priority level for mvcc gc work

irfansharif opened the pull request:
GC could be expected to be LowPri, so that it does not impact
user-facing traffic when resources (e.g. CPU, write capacity of the
store) are scarce. However long delays in GC can slow down user-facing
traffic due to more versions in the store, and can increase write
amplification of the store since there is more live data. Ideally, we
should adjust this priority based on how far behind we are with respect
to GC-ing a particular range. Keeping the priority level static at
NormalPri proved disruptive when a large volume of MVCC GC work is
suddenly accrued (if an old protected timestamp record was just released
for ex. following a long paused backup job being completed/canceled, or
just an old, long running backup job finishing).

After dynamic priority adjustment, it's not yet clear whether we need
additional pacing mechanisms to provide better latency isolation,
similar to ongoing work for backups. MVCC GC work is CPU intensive:
\#82955. This patch is also speculative in nature and in response to
observed incidents where NormalPri proved too disruptive. Fuller
treatment would entail working off of reliable reproductions of this
behaviour.

We also added a cluster setting (kv.mvcc_gc.queue_interval)
that controls how long the MVCC GC queue waits between
processing replicas. It was previously hardcoded to 1s (which is the
default value), but changing it would've come in handy in support
incidents as a form of manual pacing of MVCC GC work (we have a similar
useful knob for the merge queue).

Release note (performance improvement): Previously if there was sudden
increase in the volume of pending MVCC GC work, there was an impact on
foreground latencies. These sudden increases commonly occurred when:
- gc.ttlseconds was reduced dramatically over tables/indexes that accrue
  a lot of MVCC garbage (think "rows being frequently deleted")
- a paused backup job from a while ago (think > 1 day) was
  canceled/failed
- a backup job that started a while ago (think > 1 day) just finished

Indicators of a large increase in the volume of pending MVCC GC work is
a steep climb in the "GC Queue" graph found in the DB console page, when
navigating to 'Metrics', and selecting the 'Queues' dashboard. With this
patch, the effect on foreground latencies as a result of this sudden
build up should be reduced.

irfansharif committed:
kvserver: lower priority level for mvcc gc work

GC could be expected to be LowPri, so that it does not impact
user-facing traffic when resources (e.g. CPU, write capacity of the
store) are scarce. However long delays in GC can slow down user-facing
traffic due to more versions in the store, and can increase write
amplification of the store since there is more live data. Ideally, we
should adjust this priority based on how far behind we are with respect
to GC-ing a particular range. Keeping the priority level static at
NormalPri proved disruptive when a large volume of MVCC GC work is
suddenly accrued (if an old protected timestamp record was just released
for ex. following a long paused backup job being completed/canceled, or
just an old, long running backup job finishing).

After dynamic priority adjustment, it's not yet clear whether we need
additional pacing mechanisms to provide better latency isolation,
similar to ongoing work for backups. MVCC GC work is CPU intensive:
\#82955. This patch is also speculative in nature and in response to
observed incidents where NormalPri proved too disruptive. Fuller
treatment would entail working off of reliable reproductions of this
behaviour.

Release note (performance improvement): Previously if there was sudden
increase in the volume of pending MVCC GC work, there was an impact on
foreground latencies. These sudden increases commonly occurred when:
- gc.ttlseconds was reduced dramatically over tables/indexes that accrue
  a lot of MVCC garbage (think "rows being frequently deleted")
- a paused backup job from a while ago (think > 1 day) was
  canceled/failed
- a backup job that started a while ago (think > 1 day) just finished

Indicators of a large increase in the volume of pending MVCC GC work is
a steep climb in the "GC Queue" graph found in the DB console page, when
navigating to 'Metrics', and selecting the 'Queues' dashboard. With this
patch, the effect on foreground latencies as a result of this sudden
build up should be reduced.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85823)
<!-- Reviewable:end -->

lunevalex commented on pkg/kv/kvserver/gc/gc.go:
```diff
@@ -115,6 +116,20 @@ var MaxIntentKeyBytesPerCleanupBatch = settings.RegisterIntSetting(
 
+// AdmissionPriority determines the admission priority level to use for MVCC GC
+// work.
+var AdmissionPriority = settings.RegisterEnumSetting(
```
nit: I dont think a setting is necessary, but also have no objections to adding one.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Should we follow this up with a change that makes `mvccGCQueueTimerDuration` a cluster setting? We have precedent for that with `kv.range_merge.queue_interval`.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-:-N92f_8d0_ZDuPa4GbB7:b-7kuuky)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @lunevalex, @sumeerbhola, and @tbg)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92cnp11Ib6ToP2zDb-):*
I suspect that we will eventually need additional pacing in the keyspace iteration that the MVCC GC queue performs (e.g. `processReplicatedKeyRange`) to compute which keys to GC. In cases where many keys are being garbage collected, it is probably the case that we [will issue](https://github.com/cockroachdb/cockroach/blob/6fe360e7359ebaf984db6d53d3ccaf8a2ecf5b9a/pkg/kv/kvserver/gc/gc.go#L428) batches of GCRequests frequently enough for this change to serve as a form of pacing. However, for MVCC GC queue passes where few keys are actually GCed, this probably won't be enough. In those cases, we'll loop quickly through the entire range with no intermittent GC requests, which you've demonstrated elsewhere to be a CPU hog with adverse effects on goroutine scheduling latency.
___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92e8AR8uv3yN2ks6d6):*
> MVCC GC work is CPU intensive

I think it's also worth considering that a [future revision of MVCC GC](https://github.com/cockroachdb/cockroach/pull/42514) might do away with most of the CPU intensive work.  In such a world, the synchronous keyspace iteration might be the only work left in the MVCC GC queue, assuming we want accurate stats even with delayed GC during compaction.

That doesn't change anything in this PR, but it might shift how we think about the solution space around disruptive MVCC GC going forward.
___
*[`pkg/kv/kvserver/gc/gc.go` line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92W5uj959N60L0ioTq-r1-121:-N92etzJ2DLXg5aXdOpC:beuq99m) ([raw file](https://github.com/cockroachdb/cockroach/blob/09a21bf293bd28dd5e0bed095948eaf1af42eb35/pkg/kv/kvserver/gc/gc.go#L121)):*
<details><summary><i>Previously, lunevalex wrote…</i></summary><blockquote>

nit: I dont think a setting is necessary, but also have no objections to adding one. 
</blockquote></details>

I think it's a good idea to add one, given the potential for a negative feedback loop with starved MVCC GC and foreground traffic whose efficiency depends on timely MVCC GC. We might as well build the tool that we would need to combat this if it ever comes up in a support escalation.
___
*[`pkg/kv/kvserver/gc/gc.go` line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92fMQG6MMiNuJI64NI:-N92fMQG6MMiNuJI64NJ:bqyhlrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/09a21bf293bd28dd5e0bed095948eaf1af42eb35/pkg/kv/kvserver/gc/gc.go#L126)):*
> ```Go
> 	"the admission priority to use for mvcc gc work",
> 	"low_pri",
> 	map[int64]string{
> ```

Not for now, but I wouldn't be surprised if we follow this pattern in other places and decide to package it up as a known enum in the `settings` package.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85823)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @lunevalex, @sumeerbhola, and @tbg)

___
*[`pkg/kv/kvserver/gc/gc.go` line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92nsPA6Ka89wHfQ2ej:-N92nsPB8p1kx5qm52C2:bgm9gq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/09a21bf293bd28dd5e0bed095948eaf1af42eb35/pkg/kv/kvserver/gc/gc.go#L127)):*
> ```Go
> 	"low_pri",
> 	map[int64]string{
> 		int64(admissionpb.LowPri):    "low_pri",
> ```

Actually, are these the correct priority levels, and is this the correct default? We use `BulkNormalPri` for all other bulk jobs. `LowPri` doesn't seem to be used outside of the admission package.


<!-- Sent from Reviewable.io -->

irfansharif committed:
kvserver: add kv.mvcc_gc.queue_interval

Cluster setting that controls how long the MVCC GC queue waits between
processing replicas. It was previously hardcoded to 1s (which is the
default value), but changing it would've come in handy in support
incidents as a form of manual pacing of MVCC GC work (we have a similar
useful knob for the merge queue).

Release note: None

irfansharif reviewed (commented):
> Should we follow this up with a change that makes mvccGCQueueTimerDuration a cluster setting?

Sure, just added a commit here. Will merge on green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-:-N92uXg91JKwnpWiF-m0:b4hgnmv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @lunevalex, @nvanbenschoten, @sumeerbhola, and @tbg)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92cnp11Ib6ToP2zDb-):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I suspect that we will eventually need additional pacing in the keyspace iteration that the MVCC GC queue performs (e.g. `processReplicatedKeyRange`) to compute which keys to GC. In cases where many keys are being garbage collected, it is probably the case that we [will issue](https://github.com/cockroachdb/cockroach/blob/6fe360e7359ebaf984db6d53d3ccaf8a2ecf5b9a/pkg/kv/kvserver/gc/gc.go#L428) batches of GCRequests frequently enough for this change to serve as a form of pacing. However, for MVCC GC queue passes where few keys are actually GCed, this probably won't be enough. In those cases, we'll loop quickly through the entire range with no intermittent GC requests, which you've demonstrated elsewhere to be a CPU hog with adverse effects on goroutine scheduling latency.
</blockquote></details>

Agreed, more experimentation here will tell us. Added a sentence below
___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92e8AR8uv3yN2ks6d6):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> MVCC GC work is CPU intensive

I think it's also worth considering that a [future revision of MVCC GC](https://github.com/cockroachdb/cockroach/pull/42514) might do away with most of the CPU intensive work.  In such a world, the synchronous keyspace iteration might be the only work left in the MVCC GC queue, assuming we want accurate stats even with delayed GC during compaction.

That doesn't change anything in this PR, but it might shift how we think about the solution space around disruptive MVCC GC going forward.
</blockquote></details>

Doing MVCC GC as part of compactions makes a lot of sense (I need to periodically sift through X-noremind PRs to dig up such things). Added a sentence.
___
*[`pkg/kv/kvserver/gc/gc.go` line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92W5uj959N60L0ioTq-r1-121:-N92rBMpCaXS03cJuvmc:b-ru7w2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/09a21bf293bd28dd5e0bed095948eaf1af42eb35/pkg/kv/kvserver/gc/gc.go#L121)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think it's a good idea to add one, given the potential for a negative feedback loop with starved MVCC GC and foreground traffic whose efficiency depends on timely MVCC GC. We might as well build the tool that we would need to combat this if it ever comes up in a support escalation.
</blockquote></details>

Added it for that same reason, especially given it's something we're looking to backport to 22.1. We should want this knob to go away though after more experimentation work, if that's what you're getting at Alex.
___
*[`pkg/kv/kvserver/gc/gc.go` line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92nsPA6Ka89wHfQ2ej:-N92stQ_8HufAtdc5jYX:b-9xc142) ([raw file](https://github.com/cockroachdb/cockroach/blob/09a21bf293bd28dd5e0bed095948eaf1af42eb35/pkg/kv/kvserver/gc/gc.go#L127)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Actually, are these the correct priority levels, and is this the correct default? We use `BulkNormalPri` for all other bulk jobs. `LowPri` doesn't seem to be used outside of the admission package.
</blockquote></details>

LowPri was mentioned in an earlier revision of a comment, but I think BulkNormalPri wasn't a thing then. Changed, and removed LowPri. The important thing is to have something lower than NormalPri.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85823)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @sumeerbhola, and @tbg)

___
*[`pkg/kv/kvserver/mvcc_gc_queue.go` line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N93-JUlC1zVkGdTTmVX:-N93-JUlC1zVkGdTTmVY:b-g1x0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0d98b5567d9b6713be707a5fc70ecb8f9ed8f0/pkg/kv/kvserver/mvcc_gc_queue.go#L72)):*
> ```Go
> var mvccGCQueueInterval = settings.RegisterDurationSetting(
> 	settings.SystemOnly,
> 	"kv.mvcc_gc_merge.queue_interval",
> ```

`s/kv.mvcc_gc_merge.queue_interval/kv.mvcc_gc.queue_interval/`
___
*[`pkg/kv/kvserver/gc/gc.go` line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N92nsPA6Ka89wHfQ2ej:-N93-Rdn8PpemDtITT8Z:b-kr0tsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/09a21bf293bd28dd5e0bed095948eaf1af42eb35/pkg/kv/kvserver/gc/gc.go#L127)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

LowPri was mentioned in an earlier revision of a comment, but I think BulkNormalPri wasn't a thing then. Changed, and removed LowPri. The important thing is to have something lower than NormalPri.
</blockquote></details>

Should `HighPri` also be changed to `UserHighPri`?


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-:-N93Ma7y9keSSaMDc7iT:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten, @sumeerbhola, and @tbg)

___
*[`pkg/kv/kvserver/mvcc_gc_queue.go` line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N93-JUlC1zVkGdTTmVX:-N93MVmg7Qv0cnk5Z7yB:b-fn8mpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0d98b5567d9b6713be707a5fc70ecb8f9ed8f0/pkg/kv/kvserver/mvcc_gc_queue.go#L72)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/kv.mvcc_gc_merge.queue_interval/kv.mvcc_gc.queue_interval/`
</blockquote></details>

:facepalm: 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
> Reviewed 1 of 3 files at r1, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85823)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @nvanbenschoten, @sumeerbhola, and @tbg)

___
*[`pkg/kv/kvserver/mvcc_gc_queue.go` line 465 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85823#-N93oX7l2oVRc9kvk0p-:-N93oX7l2oVRc9kvk0p0:bhlhy0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5e7e0a0791f81ae70b41a53332c020943638c44/pkg/kv/kvserver/mvcc_gc_queue.go#L465)):*
> ```Go
> 			// BulkNormalPri.
> 			//
> 			// After we implement dynamic priority adjustment, it's not clear
> ```

It is good to have this knob to start with. But I hope we never have to tune it, or do dynamic priority adjustment in the code (it will just make things harder to understand). If someone has not provisioned well enough for GC to function, then it is their problem to add more resources.


<!-- Sent from Reviewable.io -->

irfansharif commented:
bors r-

irfansharif commented:
bors r+
